File: alpaca-auto-trading-app/backend/app/services/alpaca.py
from alpaca_trade_api.rest import REST, TimeFrame
import os

class AlpacaService:
    def __init__(self):
        self.api_key = os.getenv("APCA_API_KEY_ID")
        self.secret_key = os.getenv("APCA_API_SECRET_KEY")
        self.base_url = os.getenv("APCA_API_BASE_URL", "https://paper-api.alpaca.markets")
        self.api = REST(self.api_key, self.secret_key, self.base_url, api_version='v2')

    def get_account(self):
        return self.api.get_account()

    def get_positions(self):
        return self.api.list_positions()

    def place_order(self, symbol, qty, side, order_type='market', time_in_force='gtc'):
        return self.api.submit_order(
            symbol=symbol,
            qty=qty,
            side=side,
            type=order_type,
            time_in_force=time_in_force
        )

    def get_market_data(self, symbol, timeframe=TimeFrame.Day, limit=100):
        return self.api.get_barset(symbol, timeframe, limit=limit).df[symbol]

    def get_trade_history(self, symbol):
        return self.api.list_orders(status='all', limit=100, symbol=symbol)

File: alpaca-auto-trading-app/src/lib/api/alpaca.ts
/**
 * File: src/lib/api/alpaca.ts
 * Description: Alpaca API client and trading functions.
 */
import { OrderRequest } from '@/lib/validation';

export interface AlpacaConfig {
  apiKey: string;
  secretKey: string;
  baseUrl: string;
  paper: boolean;
}

export class AlpacaClient {
  private config: AlpacaConfig;

  constructor(config: AlpacaConfig) {
    this.config = config;
  }

  async getAccount() {
    const response = await fetch(`${this.config.baseUrl}/v2/account`, {
      headers: {
        'APCA-API-KEY-ID': this.config.apiKey,
        'APCA-API-SECRET-KEY': this.config.secretKey,
      },
    });
    return response.json();
  }

  async getPositions() {
    const response = await fetch(`${this.config.baseUrl}/v2/positions`, {
      headers: {
        'APCA-API-KEY-ID': this.config.apiKey,
        'APCA-API-SECRET-KEY': this.config.secretKey,
      },
    });
    return response.json();
  }

  async getOrders() {
    const response = await fetch(`${this.config.baseUrl}/v2/orders`, {
      headers: {
        'APCA-API-KEY-ID': this.config.apiKey,
        'APCA-API-SECRET-KEY': this.config.secretKey,
      },
    });
    return response.json();
  }

  async createOrder(order: OrderRequest) {
    const response = await fetch(`${this.config.baseUrl}/v2/orders`, {
      method: 'POST',
      headers: {
        'APCA-API-KEY-ID': this.config.apiKey,
        'APCA-API-SECRET-KEY': this.config.secretKey,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(order),
    });
    return response.json();
  }

  async cancelOrder(orderId: string) {
    const response = await fetch(`${this.config.baseUrl}/v2/orders/${orderId}`, {
      method: 'DELETE',
      headers: {
        'APCA-API-KEY-ID': this.config.apiKey,
        'APCA-API-SECRET-KEY': this.config.secretKey,
      },
    });
    return response.json();
  }
}

export const alpacaClient = new AlpacaClient({
  apiKey: process.env.APCA_API_KEY_ID || '',
  secretKey: process.env.APCA_API_SECRET_KEY || '',
  baseUrl: process.env.APCA_PAPER_BASE_URL || 'https://paper-api.alpaca.markets',
  paper: process.env.NODE_ENV !== 'production',
});

export async function alpacaTrade(symbol: string, quantity: number, side: 'buy' | 'sell') {
  return alpacaClient.createOrder({
    symbol,
    qty: quantity,
    side,
    type: 'market',
    time_in_force: 'day',
  });
}

export function getAlpacaService() {
  return alpacaClient;
}